📈 feat: Chat rating for feedback

rubentalstra opened the pull request:
## Summary

- Closes: #901

This pull request introduces a new feature for providing feedback on messages, along with various enhancements to the user interface and backend. The most important changes include adding a feedback route to the server, creating a new component for feedback tag options, and updating the UI to handle feedback submission.

### Backend changes:
* [`api/server/routes/messages.js`](diffhunk://#diff-b0de6591908ddf616e0416112ece7521a332c3f4840410b3c128f5193c912b4bR178-R236): Added a new PUT route for updating message feedback, including handling thumbs down feedback with default tag choices.

### Frontend changes:
* [`client/src/components/Chat/Messages/FeedbackTagOptions.tsx`](diffhunk://#diff-4cb9b41bccc1444d3e4c1600e21e818b2229dab319d34a842143d6401c8df06dR1-R150): Created a new `FeedbackTagOptions` component to allow users to select feedback tags and provide additional comments.
* [`client/src/components/Chat/Messages/HoverButtons.tsx`](diffhunk://#diff-b73e9e93939615fec9fd27fe78a52c5c59f65b2ad4b6185511329c61e7feea9cL3-R12): Updated the `HoverButtons` component to include thumbs up and thumbs down buttons for message feedback, and handle feedback submission state. [[1]](diffhunk://#diff-b73e9e93939615fec9fd27fe78a52c5c59f65b2ad4b6185511329c61e7feea9cL3-R12) [[2]](diffhunk://#diff-b73e9e93939615fec9fd27fe78a52c5c59f65b2ad4b6185511329c61e7feea9cR31-R32) [[3]](diffhunk://#diff-b73e9e93939615fec9fd27fe78a52c5c59f65b2ad4b6185511329c61e7feea9cR47-R48) [[4]](diffhunk://#diff-b73e9e93939615fec9fd27fe78a52c5c59f65b2ad4b6185511329c61e7feea9cR79-R84) [[5]](diffhunk://#diff-b73e9e93939615fec9fd27fe78a52c5c59f65b2ad4b6185511329c61e7feea9cR135-R165)
* [`client/src/components/Chat/Messages/ui/MessageRender.tsx`](diffhunk://#diff-fff7b0e8a3c010748ac80aef0f8d2d6ad8329b6cd622c9e74b7af05893552871R1-R5): Integrated the `FeedbackTagOptions` component into the `MessageRender` component to display feedback options when a message is rated thumbs down. [[1]](diffhunk://#diff-fff7b0e8a3c010748ac80aef0f8d2d6ad8329b6cd622c9e74b7af05893552871R1-R5) [[2]](diffhunk://#diff-fff7b0e8a3c010748ac80aef0f8d2d6ad8329b6cd622c9e74b7af05893552871R54-R55) [[3]](diffhunk://#diff-fff7b0e8a3c010748ac80aef0f8d2d6ad8329b6cd622c9e74b7af05893552871R64-R67) [[4]](diffhunk://#diff-fff7b0e8a3c010748ac80aef0f8d2d6ad8329b6cd622c9e74b7af05893552871R215-R244)

### New icons:
* [`client/src/components/svg/ThumbDownIcon.tsx`](diffhunk://#diff-33a6c9aa80c5daed0ae803d3f9ff6dc379cef5df5db930a05ef134ab12232025R1-R35): Added a new `ThumbDownIcon` component for the thumbs down feedback button.
* [`client/src/components/svg/ThumbUpIcon.tsx`](diffhunk://#diff-a779527e2eff21c911afd773209406b1ff60a9e1a168808dc04bf694e37e3727R1-R35): Added a new `ThumbUpIcon` component for the thumbs up feedback button.
* [`client/src/components/svg/index.ts`](diffhunk://#diff-cff1a9be7379b419c365b01d97f7de369f6fe42393a9839fa940d8610a9cf5e7R59-R60): Exported the new `ThumbUpIcon` and `ThumbDownIcon` components.

### Hook updates:
* [`client/src/hooks/Messages/useMessageActions.tsx`](diffhunk://#diff-6ae562d4c51b40d76a6b4fa72a93329b1f2e82323cab55d25eaf9276a02ea16bL2-R8): Updated the `useMessageActions` hook to handle feedback submission and state management for message ratings. [[1]](diffhunk://#diff-6ae562d4c51b40d76a6b4fa72a93329b1f2e82323cab55d25eaf9276a02ea16bL2-R8) [[2]](diffhunk://#diff-6ae562d4c51b40d76a6b4fa72a93329b1f2e82323cab55d25eaf9276a02ea16bR20) [[3]](diffhunk://#diff-6ae562d4c51b40d76a6b4fa72a93329b1f2e82323cab55d25eaf9276a02ea16bR55)


## Change Type

- [x] New feature (non-breaking change which adds functionality)
- [x] Translation update

## Testing

### **Test Configuration:**
- **Backend:**  
  - Tested the new `/feedback` API endpoint using Postman to ensure proper handling of feedback submissions, including defaulting tag choices for thumbs-down ratings.
  
- **Frontend:**  
  - Verified that clicking the thumbs up/down buttons triggers the correct feedback flow.
  - Confirmed that when a user submits a thumbs down rating, the `FeedbackTagOptions` component renders with predefined tags.
  - Ensured that after a tag is selected, a thank-you message appears for a few seconds.
  - Conducted local integration testing to verify UI updates and mutation flows.

- **General:**  
  - All local unit tests pass without errors.
  - No new warnings were introduced.

## Checklist

- [x] My code adheres to this project's style guidelines.
- [x] I have performed a self-review of my own code.
- [x] I have commented on complex sections of the code.
- [x] I have updated relevant documentation.
- [x] My changes do not introduce new warnings.
- [x] Local unit tests pass with my changes.
- [x] Dependent changes have been merged and published in downstream modules.

rubentalstra committed:
feat: working started for feedback implementation.

TODO:
- needs some refactoring.
- needs some UI animations.

rubentalstra commented:
<img width="1381" alt="working_example_screenshot" src="https://github.com/user-attachments/assets/d32e4aa2-87f3-499a-ab9c-9596a663c2d5" />

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
feat: working rate functionality

rubentalstra committed:
feat: works now as well to reader the already rated responses from the server.

rubentalstra commented:
<img width="797" alt="Screenshot 2025-02-15 at 14 44 01" src="https://github.com/user-attachments/assets/88c6f67b-3dbd-4e80-8b57-9bb6aa35747f" />
<img width="797" alt="Screenshot 2025-02-15 at 14 44 24" src="https://github.com/user-attachments/assets/42a13f23-cb9f-4b1f-a153-452dd8e30941" />
<img width="797" alt="Screenshot 2025-02-15 at 14 44 38" src="https://github.com/user-attachments/assets/874862ff-74ff-45d9-aced-3590db246425" />
<img width="797" alt="Screenshot 2025-02-15 at 14 44 43" src="https://github.com/user-attachments/assets/ce57b130-6299-451c-856d-ff1e01235dcb" />

rubentalstra committed:
feat: added the option to give feedback in text (optional)

rubentalstra committed:
feat: added Dismiss option `x` to the `FeedbackTagOptions`

walbercardoso commented:
Nice work. Is the feedback I send stored somewhere?

rubentalstra commented:
@walbercardoso yes this is stored on the message object.

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
✨ feat: Add rating and ratingContent fields to message schema

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
🔧 chore: Bump version to 0.0.3 in package.json

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra commented:
@danny-avila, it works as is. But if you want to have a better looking UI then @berry-13 needs to do his magic.

But we can also merge it first and then improve UI design. It's up to you.

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
✨ feat: Enhance feedback localization and update UI elements

rubentalstra commented:
@danny-avila, there is only a problem with agents, maybe you can have a look why this is the case only for conversations with agents?

```bash
TypeError: handleFeedback is not a function. (In 'handleFeedback("thumbsDown")', 'handleFeedback' is undefined)
TypeError: handleFeedback is not a function. (In 'handleFeedback("thumbsUp")', 'handleFeedback' is undefined)
```

benverhees commented:
@rubentalstra I see that all response messages get a `rating` and a `ratingContent` with all tagChoices. Is that intentional?
```
    rating: null,
    ratingContent: {
        tagChoices: [
            'Shouldn\'t have used Memory',
            'Don\'t like the style',
            'Not factually correct',
            'Didn\'t fully follow instructions',
            'Refused when it shouldn\'t have',
            'Being lazy',
            'Unsafe or problematic',
            'Biased',
            'Other'
        ],
        tags: [],
        text: null
    },
```
Also, the choises are lacking localization. The choices would better be stored using a key and having translations shown in the UI. I can have a look at both these improvements, I've got something working locally.

rubentalstra commented:
@benverhees Hi, thank your for having a look at this PR,

I just reversed engineered the OpenIA way of doing this feedback message stuff.

This was just a first attempt of implementation. But feedback on this is always welcome 🤗

benverhees commented:
> @danny-avila, there is only a problem with agents, maybe you can have a look why this is the case only for conversations with agents?
> 
> ```shell
> TypeError: handleFeedback is not a function. (In 'handleFeedback("thumbsDown")', 'handleFeedback' is undefined)
> TypeError: handleFeedback is not a function. (In 'handleFeedback("thumbsUp")', 'handleFeedback' is undefined)
> ```

While refactoring, I also found why this was happening; agent messages are rendered by a different component that didn't pass the `handleFeedback` function. My update will address this as well.

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
🚀 feat: Implement feedback tagging system with thumbs up/down options

rubentalstra committed:
🚀 feat: Add data-provider package to unused i18n keys detection

berry-13 committed:
🎨 style: update HoverButtons' style

berry-13 committed:
🎨 style: Update HoverButtons and Fork components for improved styling and visibility

benverhees commented:
@rubentalstra @berry-13 I've got an updated version of the chat rating in https://github.com/benverhees/LibreChat/tree/feat/messages-feedback
It has a new styling for the feedback panel, and added functionality, supporting feedback options for positive feedback, updating and removing given feedback.

rubentalstra commented:
> @rubentalstra @berry-13 I've got an updated version of the chat rating in https://github.com/benverhees/LibreChat/tree/feat/messages-feedback It has a new styling for the feedback panel, and added functionality, supporting feedback options for positive feedback, updating and removing given feedback.

@benverhees thank you. could you please compare it to this PR. I have updated it also in the meantime. and please use the `thumpsDown` and `thumpsUp`. because positive and negative I personally don't like. (that's a me issue 😉)

and thank you again for the time. we will take the changes into review and see what we are going to see. (remove the feedback was still on the TODO list, and thank you for that.)

benverhees commented:
@rubentalstra I've rebased my changes onto the current state of this branch and using `thumbsUp` and `thumbsDown` everywhere now. 👍 👎

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
🔧 feat: Implement feedback system with rating and content options

rubentalstra committed:
🔧 feat: Enhance feedback handling with improved rating toggle and tag options

rubentalstra committed:
🔧 feat: Integrate toast notifications for feedback submission and clean up unused state

rubentalstra committed:
🔧 feat: Remove unused feedback tag options from translation file

berry-13 committed:
Merge branch 'main' into feat/messages-feedback

berry-13 committed:
✨ refactor: clean up Feedback component and improve HoverButtons structure

berry-13 committed:
✨ refactor: remove unused settings switches for auto scroll, hide side panel, and user message markdown

berry-13 committed:
refactor: reorganize import order

berry-13 committed:
✨ refactor: enhance HoverButtons and Fork components with improved styles and animations

berry-13 committed:
✨ refactor: update feedback response phrases for improved user engagement

berry-13 committed:
✨ refactor: add CheckboxOption component and streamline fork options rendering

berry-13 committed:
Refactor feedback components and logic

- Consolidated feedback handling into a single Feedback component, removing FeedbackButtons and FeedbackTagOptions.
- Introduced new feedback tagging system with detailed tags for both thumbs up and thumbs down ratings.
- Updated feedback schema to include new tags and improved type definitions.
- Enhanced user interface for feedback collection, including a dialog for additional comments.
- Removed obsolete files and adjusted imports accordingly.
- Updated translations for new feedback tags and placeholders.

berry-13 committed:
✨ refactor: update feedback handling by replacing rating fields with feedback in message updates

berry-13 committed:
fix: add missing validateMessageReq middleware to feedback route and refactor feedback system

berry-13 commented:
> @berry-13 merge conflicts. can you fix them? if not, I will fix them by end of day today (in roughly 12-15 hours for me) as I want it in dev branch

I'll fix them in 2/3 hours

rubentalstra committed:
Merge branch 'main' into feat/messages-feedback

rubentalstra committed:
🗑️ chore: Remove redundant fork option explanations from translation file

rubentalstra committed:
🔧 refactor: Remove unused dependency from feedback callback

rubentalstra committed:
🔧 refactor: Simplify message update response structure and improve error logging

rubentalstra committed:
Chore: removed unused tests.

rubentalstra committed:
Merge branch 'dev' into feat/messages-feedback

JIPINGBABA commented:
Could you please tell me how to use the Chat rating for feedback feature,where can I get this version,thanks so much

Zafnok commented:
Thanks for this feature. I was wondering if there's a non-hacky way supported to have it route the feedback to a custom endpoint for processing? i.e. a setting in librechat.yaml to specify some feedback endpoint so we can collect good training data and filter the bad. I looked in the docs but I suppose since it's a new feature it's not heavily documented. It would be great if it was supported out of the box without having to do some MongoDB hook voodoo.
